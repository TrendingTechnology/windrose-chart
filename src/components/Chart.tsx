/* eslint-disable react/no-this-in-sfc */
/* eslint-disable func-names */
import * as React from "react";
import * as d3 from "d3";
import styled from "styled-components";
import { debounce } from "lodash";
import { DefaultProps } from "./DefaultProps";
import { PropType, DataType } from "./Types";

export function Chart(props: PropType): JSX.Element {
  const {
    width,
    height,
    data,
    columns,
    columnsColor: z,
    angles,
    dataMax,
    dataKeys,
    mouseOverColor,
    mouseOverTitleColor,
    mouseOverSurveyColor,
  } = props;
  const containerRef = React.useRef<SVGSVGElement>(null);
  const axisContainerRef = React.useRef<HTMLDivElement>(null);
  React.useEffect(() => {
    const svg = d3.select(containerRef.current);
    svg.selectAll("*").remove();
    const margin = { top: 80, right: 100, bottom: 80, left: 40 };
    const innerRadius = 20;
    const chartWidth = width - margin.left - margin.right;
    const chartHeight = height - margin.top - margin.bottom;
    const outerRadius =
      innerRadius + Math.min(chartWidth, chartHeight) / 2;
    const g = svg
      .append("g")
      .attr("transform", `translate(${width / 2}, ${height / 2})`);
    const angle = d3.scaleLinear().range([0, 2 * Math.PI]);
    const radius = d3.scaleLinear().range([innerRadius, outerRadius]);
    const x = d3
      .scaleBand()
      .range([0, 2 * Math.PI])
      .align(0);
    const xGroup = d3
      .scaleBand()
      .range([0, 2 * Math.PI])
      .align(0);
    const y = d3.scaleLinear().range([innerRadius, outerRadius]);
    x.domain(angles.map((d) => d));
    xGroup.domain(columns.map((d) => d));
    y.domain([
      0,
      (d3.max(data, ({ total }) => total) as number) > dataMax
        ? (d3.max(data, ({ total }) => total) as number)
        : dataMax,
    ]);
    angle.domain([
      0,
      d3.max(data, (_, i): number => i + 1) as number,
    ]);
    radius.domain([0, d3.max(data, () => 0) as number]);
    const angleOffset = -360.0 / data.length / 2.0;
    const stackGen: d3.Stack<any, DataType, string> = d3
      .stack()
      .keys(dataKeys);
    const arcVal: d3.Arc<SVGPathElement, d3.DefaultArcObject> = d3
      .arc()
      .innerRadius((d) => Number(y(d[0])))
      .outerRadius((d) => Number(y(d[1])))
      .startAngle((_d, i) => Number(x(angles[i])))
      .endAngle((_d, i) => Number(x(angles[i])) + x.bandwidth())
      .padAngle(0.0)
      .padRadius(innerRadius);
    const arcParent = g
      .append("g")
      .selectAll("g")
      .data(stackGen(data))
      .enter();
    const arc: d3.Selection<
      SVGPathElement,
      d3.SeriesPoint<DataType>,
      d3.EnterElement,
      d3.Series<DataType, string>
    > = arcParent
      .selectAll<SVGPathElement, d3.SeriesPoint<DataType>>("path")
      .data((d) => d)
      .enter()
      .append("path");
    arc
      // @ts-ignore
      .attr("d", arcVal)
      .attr("transform", `rotate(${angleOffset})`)
      .attr("fill", (d: d3.SeriesPoint<DataType>) => d.data.color)
      .attr(
        "class",
        (_d: d3.SeriesPoint<DataType>, i: number) => `arc_${i}`
      )
      .on(
        "mouseover",
        debounce(function (
          this: SVGPathElement,
          _event: any,
          d: d3.SeriesPoint<DataType>
        ) {
          const classname = this.getAttribute("class") || "arc_1";
          const i = classname.split("_")[1];
          this.setAttribute("fill", mouseOverColor);
          this.setAttribute(
            "style",
            "transition: fill 0.5s; cursor: pointer;"
          );
          const bbox = this.getBBox();
          arcParent
            .append("text")
            .attr("class", `tooltip_${i}`)
            .attr("x", bbox.x + bbox.width / 2)
            .attr("dy", bbox.y + bbox.height / 2)
            .attr("text-anchor", "middle")
            .style("font-size", 12)
            .style("fill", mouseOverTitleColor)
            .style("opacity", 1)
            .style("transition", "opacity 0.2s")
            .attr("font-weight", 600)
            .text(`${d.data.coreCompetency}`);
          arcParent
            .append("text")
            .attr("class", `tooltip_${i}`)
            .attr("x", bbox.x + bbox.width / 2)
            .attr("dy", bbox.y + bbox.height / 2 + 14)
            .attr("text-anchor", "middle")
            .style("font-size", 14)
            .style("fill", mouseOverSurveyColor)
            .style("opacity", 1)
            .style("transition", "opacity 0.2s")
            .attr("font-weight", 800)
            .text(`${d.data.survey}`);
        },
        0)
      )
      .on(
        "mouseout",
        debounce(function (
          this: SVGPathElement,
          _event: any,
          d: d3.SeriesPoint<DataType>
        ) {
          const classname = this.getAttribute("class") || "arc_1";
          const i = classname.split("_")[1];
          this.setAttribute(
            "fill",
            String(d.data!.color) || "#ffffff"
          );
          // this.setAttribute("style", "opacity: 0");
          d3.selectAll(`.tooltip_${i}`).remove();
        },
        0)
      );

    const label = g
      .append("g")
      .selectAll("g")
      .data(columns)
      .enter()
      .append("g")
      .attr("text-anchor", "middle")
      .attr("transform", (d) => {
        if (typeof d === "undefined") return null;
        return `rotate(${
          ((Number(xGroup(d)) + xGroup.bandwidth() / 2) * 180) /
            Math.PI -
          (90 - angleOffset)
        })translate(${outerRadius},0)`;
      });
    label
      .append("text")
      .attr("transform", "rotate(90)translate(0,-9)")
      .text((_d, i) => columns[i])
      .style("font-size", 14);
    const yAxis = g.append("g").attr("text-anchor", "middle");
    const yTick = yAxis
      .selectAll("g")
      .data(y.ticks(5).slice(1))
      .enter()
      .append("g");
    yTick
      .append("circle")
      .attr("fill", "none")
      .attr("stroke", "gray")
      .attr("stroke-dasharray", "4,4")
      // @ts-ignore
      .attr("r", y);
    yTick
      .append("text")
      // @ts-ignore
      .attr("y", (d) => -y(d))
      .attr("dy", "-0.35em")
      .attr("x", () => -10)
      .text(y.tickFormat(5, "s"))
      .style("font-size", 14);
    const legend = g
      .append("g")
      .selectAll("g")
      .data(columns)
      .enter()
      .append("g")
      .attr(
        "transform",
        (_d, i) =>
          `translate(${outerRadius + 0},${
            -outerRadius + 40 + (i - (columns.length - 1) / 2) * 20
          })`
      );
    legend
      .append("rect")
      .attr("width", 18)
      .attr("height", 18)
      .attr("fill", (_d, i) => z[i]);
    legend
      .append("text")
      .attr("x", 24)
      .attr("y", 9)
      .attr("dy", "0.35em")
      .text((d) => d)
      .style("font-size", 12);
    g.exit().remove();
  });
  return (
    <AxisContainer ref={axisContainerRef} role="main">
      <Axis
        className="axis"
        width={width}
        height={height}
        ref={containerRef}
        role="document"
      />
    </AxisContainer>
  );
}

export const AxisContainer = styled.div`
  position: relative;
`;

export const Axis = styled.svg`
  .axis {
    position: absolute;
    top: 0;
    left: 0;
    stroke: gray;
  }
`;

const {
  data,
  angles,
  columns,
  columnsColor,
  width,
  height,
  dataMax,
  dataKeys,
  mouseOverColor,
  mouseOverTitleColor,
  mouseOverSurveyColor,
} = DefaultProps;

Chart.defaultProps = {
  data,
  dataMax,
  angles,
  columns,
  columnsColor,
  width,
  height,
  dataKeys,
  mouseOverColor,
  mouseOverTitleColor,
  mouseOverSurveyColor,
};

export default Chart;
