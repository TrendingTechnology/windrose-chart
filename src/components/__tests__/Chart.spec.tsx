import * as React from "react";
import { render } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";
import Chart from "../../index";
import "jest-styled-components";

describe("<Chart />", () => {
  it("matches snapshot", () => {
    const utils = render(<Chart />);
    expect(utils.container).toMatchSnapshot();
  });
});

// test("Renders", async () => {
//   const {  } = render(<Chart />);
//   expect(getByRole("document")).toBeValid();
// });
