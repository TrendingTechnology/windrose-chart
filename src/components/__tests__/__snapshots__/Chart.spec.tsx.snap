// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Chart /> matches snapshot 1`] = `
.c0 {
  position: relative;
}

.c1 .axis {
  position: absolute;
  top: 0;
  left: 0;
  stroke: gray;
}

<div>
  <div
    class="c0"
    role="main"
  >
    <svg
      class="c1 axis"
      height="500"
      role="document"
      width="600"
    >
      <g
        transform="translate(300, 250)"
      >
        <g>
          <path
            class="arc_0"
            d="M3.3418327908936248e-15,-54.57627118644068A54.57627118644068,54.57627118644068,0,0,1,11.347044820901449,-53.38364871801448L4.158233816355189,-19.56295201467611A20,20,0,0,0,1.2246467991473533e-15,-20Z"
            fill="#8e44ad"
            transform="rotate(-6)"
          />
          <path
            class="arc_1"
            d="M17.936788241735517,-84.38595402940796A86.27118644067795,86.27118644067795,0,0,1,35.089652767047845,-78.81265049831929L8.134732861516005,-18.270909152852017A20,20,0,0,0,4.158233816355189,-19.56295201467611Z"
            fill="#7e44ad"
            transform="rotate(-6)"
          />
          <path
            class="arc_2"
            d="M36.2616058064188,-81.44490012203526A89.15254237288136,89.15254237288136,0,0,1,52.40254961115947,-72.12592187139361L11.755705045849464,-16.18033988749895A20,20,0,0,0,8.134732861516005,-18.270909152852017Z"
            fill="#6e44ad"
            transform="rotate(-6)"
          />
          <path
            class="arc_3"
            d="M64.25787927604155,-88.44338328336289A109.32203389830508,109.32203389830508,0,0,1,81.2421038021897,-73.15071883075653L14.862896509547884,-13.382612127177165A20,20,0,0,0,11.755705045849464,-16.18033988749895Z"
            fill="#5e44ad"
            transform="rotate(-6)"
          />
          <path
            class="arc_4"
            d="M51.26439728293211,-46.15867064204327A68.98305084745762,68.98305084745762,0,0,1,59.74107446445194,-34.49152542372881L17.32050807568877,-10A20,20,0,0,0,14.862896509547884,-13.382612127177165Z"
            fill="#4e44ad"
            transform="rotate(-6)"
          />
          <path
            class="arc_5"
            d="M122.12426033028015,-70.50847457627118A141.01694915254237,141.01694915254237,0,0,1,134.11508839958773,-43.57663378304343L19.02113032590307,-6.180339887498948A20,20,0,0,0,17.32050807568877,-10Z"
            fill="#c0392b"
            transform="rotate(-6)"
          />
          <path
            class="arc_6"
            d="M172.47974109081596,-56.04206508155825A181.3559322033898,181.3559322033898,0,0,1,180.3624454311953,-18.95685689769307L19.890437907365467,-2.090569265353068A20,20,0,0,0,19.02113032590307,-6.180339887498948Z"
            fill="#c0492b"
            transform="rotate(-6)"
          />
          <path
            class="arc_7"
            d="M65.739582914174,-6.909508588878797A66.10169491525423,66.10169491525423,0,0,1,65.739582914174,6.909508588878782L19.890437907365467,2.090569265353068A20,20,0,0,0,19.890437907365467,-2.0905692653530723Z"
            fill="#c0592b"
            transform="rotate(-6)"
          />
          <path
            class="arc_8"
            d="M85.79858385465272,9.017794542921282A86.27118644067795,86.27118644067795,0,0,1,82.04877403292086,26.659262735059016L19.02113032590307,6.180339887498948A20,20,0,0,0,19.890437907365467,2.090569265353068Z"
            fill="#c0692b"
            transform="rotate(-6)"
          />
          <path
            class="arc_9"
            d="M172.47974109081596,56.04206508155825A181.3559322033898,181.3559322033898,0,0,1,157.05884441514397,90.67796610169488L17.320508075688775,9.999999999999996A20,20,0,0,0,19.02113032590307,6.180339887498948Z"
            fill="#c0792b"
            transform="rotate(-6)"
          />
          <path
            class="arc_10"
            d="M162.04949928441022,93.55932203389827A187.1186440677966,187.1186440677966,0,0,1,139.0562520893294,125.20681176613206L14.86289650954789,13.382612127177161A20,20,0,0,0,17.320508075688775,9.999999999999996Z"
            fill="#27ae60"
            transform="rotate(-6)"
          />
          <path
            class="arc_11"
            d="M126.20866358107615,113.6387911138264A169.83050847457628,169.83050847457628,0,0,1,99.82386827068787,137.3957675192707L11.755705045849469,16.180339887498942A20,20,0,0,0,14.86289650954789,13.382612127177161Z"
            fill="#27ae70"
            transform="rotate(-6)"
          />
          <path
            class="arc_12"
            d="M98.13024974713322,135.06470160327513A166.94915254237287,166.94915254237287,0,0,1,67.90433786943447,152.5156399623664L8.134732861516007,18.270909152852017A20,20,0,0,0,11.755705045849464,16.18033988749895Z"
            fill="#27ae80"
            transform="rotate(-6)"
          />
          <path
            class="arc_13"
            d="M74.93605610566017,168.30913770466222A184.23728813559322,184.23728813559322,0,0,1,38.30508608794996,180.21126135553334L4.158233816355193,19.56295201467611A20,20,0,0,0,8.134732861516007,18.270909152852017Z"
            fill="#27ae90"
            transform="rotate(-6)"
          />
          <path
            class="arc_14"
            d="M29.319072332267126,137.93539047636037A141.01694915254237,141.01694915254237,0,0,1,7.125890329487831e-14,141.01694915254237L1.0106430996148605e-14,20A20,20,0,0,0,4.158233816355193,19.56295201467611Z"
            fill="#27aea0"
            transform="rotate(-6)"
          />
          <path
            class="arc_15"
            d="M3.518264956872481e-15,57.45762711864407A57.45762711864407,57.45762711864407,0,0,1,-11.946112404613618,56.20204010995935L-4.158233816355183,19.562952014676114A20,20,0,0,0,1.2246467991473533e-15,20Z"
            fill="#2c3e50"
            transform="rotate(-6)"
          />
          <path
            class="arc_16"
            d="M-33.51254541825236,157.66413021997442A161.1864406779661,161.1864406779661,0,0,1,-65.56043179069249,147.2511407149345L-8.134732861515998,18.27090915285202A20,20,0,0,0,-4.158233816355183,19.562952014676114Z"
            fill="#2c4e50"
            transform="rotate(-6)"
          />
          <path
            class="arc_17"
            d="M-22.198169333967382,49.85790463744365A54.57627118644068,54.57627118644068,0,0,1,-32.07912732850444,44.153130879446294L-11.755705045849453,16.180339887498953A20,20,0,0,0,-8.134732861515998,18.27090915285202Z"
            fill="#2c5e50"
            transform="rotate(-6)"
          />
          <path
            class="arc_18"
            d="M-50.70893108760487,69.79485595539799A86.27118644067795,86.27118644067795,0,0,1,-64.11198579118535,57.72669129434895L-14.862896509547882,13.382612127177167A20,20,0,0,0,-11.75570504584946,16.18033988749895Z"
            fill="#2c6e50"
            transform="rotate(-6)"
          />
          <path
            class="arc_19"
            d="M-38.41680877467885,34.59064998973759A51.69491525423729,51.69491525423729,0,0,1,-44.76910985665317,25.84745762711866L-17.32050807568877,10.000000000000007A20,20,0,0,0,-14.862896509547882,13.382612127177167Z"
            fill="#2c7e50"
            transform="rotate(-6)"
          />
          <path
            class="arc_20"
            d="M-157.0588444151439,90.67796610169496A181.3559322033898,181.3559322033898,0,0,1,-172.47974109081596,56.04206508155827L-19.02113032590307,6.18033988749895A20,20,0,0,0,-17.32050807568877,10.000000000000007Z"
            fill="#0984e3"
            transform="rotate(-6)"
          />
          <path
            class="arc_21"
            d="M-158.7780794153773,51.59012533208869A166.94915254237287,166.94915254237287,0,0,1,-166.03458761656768,17.450938359091268L-19.890437907365467,2.0905692653530656A20,20,0,0,0,-19.02113032590307,6.18033988749895Z"
            fill="#0974e3"
            transform="rotate(-6)"
          />
          <path
            class="arc_22"
            d="M-171.7657307424187,18.053305774532134A172.71186440677963,172.71186440677963,0,0,1,-171.7657307424187,-18.05330577453194L-19.890437907365467,-2.090569265353061A20,20,0,0,0,-19.890437907365467,2.0905692653530834Z"
            fill="#0964e3"
            transform="rotate(-6)"
          />
          <path
            class="arc_23"
            d="M-148.84115823901448,-15.643836112769094A149.66101694915255,149.66101694915255,0,0,1,-142.33608540485096,-46.24779763272517L-19.021130325903073,-6.1803398874989455A20,20,0,0,0,-19.890437907365467,-2.090569265353061Z"
            fill="#0954e3"
            transform="rotate(-6)"
          />
          <path
            class="arc_24"
            d="M-180.7007380960792,-58.71322893123998A190,190,0,0,1,-164.54482671904333,-95.00000000000001L-17.32050807568877,-10.000000000000002A20,20,0,0,0,-19.021130325903073,-6.1803398874989455Z"
            fill="#0944e3"
            transform="rotate(-6)"
          />
          <path
            class="arc_25"
            d="M-64.73172933371823,-37.37288135593216A74.7457627118644,74.7457627118644,0,0,1,-55.54692678568321,-50.014677526145135L-14.86289650954789,-13.382612127177158A20,20,0,0,0,-17.32050807568878,-9.999999999999988Z"
            fill="#e84393"
            transform="rotate(-6)"
          />
          <path
            class="arc_26"
            d="M-139.0562520893294,-125.20681176613205A187.1186440677966,187.1186440677966,0,0,1,-109.98557941201533,-151.38216301524437L-11.755705045849465,-16.180339887498945A20,20,0,0,0,-14.86289650954789,-13.382612127177158Z"
            fill="#d84393"
            transform="rotate(-6)"
          />
          <path
            class="arc_27"
            d="M-47.32169404049573,-65.13272412340677A80.50847457627118,80.50847457627118,0,0,1,-32.745746688305935,-73.54815125088736L-8.134732861516001,-18.27090915285202A20,20,0,0,0,-11.755705045849465,-16.180339887498945Z"
            fill="#c84393"
            transform="rotate(-6)"
          />
          <path
            class="arc_28"
            d="M-30.4018406095641,-68.28365200345539A74.7457627118644,74.7457627118644,0,0,1,-15.54051790688679,-73.11238846162851L-4.158233816355196,-19.56295201467611A20,20,0,0,0,-8.134732861516019,-18.27090915285201Z"
            fill="#b84393"
            transform="rotate(-6)"
          />
          <path
            class="arc_29"
            d="M-14.34238273946241,-67.47560567773878A68.98305084745762,68.98305084745762,0,0,1,-1.2671980862363712e-14,-68.98305084745762L-3.673940397442059e-15,-20A20,20,0,0,0,-4.158233816355196,-19.56295201467611Z"
            fill="#a84393"
            transform="rotate(-6)"
          />
        </g>
        <g>
          <g
            text-anchor="middle"
            transform="rotate(-66)translate(190,0)"
          >
            <text
              transform="rotate(90)translate(0,-9)"
            >
              Language Arts
            </text>
          </g>
          <g
            text-anchor="middle"
            transform="rotate(-6)translate(190,0)"
          >
            <text
              transform="rotate(90)translate(0,-9)"
            >
              Math
            </text>
          </g>
          <g
            text-anchor="middle"
            transform="rotate(53.99999999999997)translate(190,0)"
          >
            <text
              transform="rotate(90)translate(0,-9)"
            >
              Science
            </text>
          </g>
          <g
            text-anchor="middle"
            transform="rotate(113.99999999999997)translate(190,0)"
          >
            <text
              transform="rotate(90)translate(0,-9)"
            >
              Social Studies
            </text>
          </g>
          <g
            text-anchor="middle"
            transform="rotate(174)translate(190,0)"
          >
            <text
              transform="rotate(90)translate(0,-9)"
            >
              21st Century Skills
            </text>
          </g>
          <g
            text-anchor="middle"
            transform="rotate(233.99999999999994)translate(190,0)"
          >
            <text
              transform="rotate(90)translate(0,-9)"
            >
              Arts, Crafts & Labor
            </text>
          </g>
        </g>
        <g
          text-anchor="middle"
        >
          <g>
            <circle
              fill="none"
              r="48.813559322033896"
              stroke="gray"
              stroke-dasharray="4,4"
            />
            <text
              dy="-0.35em"
              x="-10"
              y="-48.813559322033896"
            >
              1
            </text>
          </g>
          <g>
            <circle
              fill="none"
              r="77.62711864406779"
              stroke="gray"
              stroke-dasharray="4,4"
            />
            <text
              dy="-0.35em"
              x="-10"
              y="-77.62711864406779"
            >
              2
            </text>
          </g>
          <g>
            <circle
              fill="none"
              r="106.4406779661017"
              stroke="gray"
              stroke-dasharray="4,4"
            />
            <text
              dy="-0.35em"
              x="-10"
              y="-106.4406779661017"
            >
              3
            </text>
          </g>
          <g>
            <circle
              fill="none"
              r="135.25423728813558"
              stroke="gray"
              stroke-dasharray="4,4"
            />
            <text
              dy="-0.35em"
              x="-10"
              y="-135.25423728813558"
            >
              4
            </text>
          </g>
          <g>
            <circle
              fill="none"
              r="164.0677966101695"
              stroke="gray"
              stroke-dasharray="4,4"
            />
            <text
              dy="-0.35em"
              x="-10"
              y="-164.0677966101695"
            >
              5
            </text>
          </g>
        </g>
        <g>
          <g
            transform="translate(190,-200)"
          >
            <rect
              fill="#8e44ad"
              height="18"
              width="18"
            />
            <text
              dy="0.35em"
              x="24"
              y="9"
            >
              Language Arts
            </text>
          </g>
          <g
            transform="translate(190,-180)"
          >
            <rect
              fill="#c0392b"
              height="18"
              width="18"
            />
            <text
              dy="0.35em"
              x="24"
              y="9"
            >
              Math
            </text>
          </g>
          <g
            transform="translate(190,-160)"
          >
            <rect
              fill="#27ae60"
              height="18"
              width="18"
            />
            <text
              dy="0.35em"
              x="24"
              y="9"
            >
              Science
            </text>
          </g>
          <g
            transform="translate(190,-140)"
          >
            <rect
              fill="#2c3e50"
              height="18"
              width="18"
            />
            <text
              dy="0.35em"
              x="24"
              y="9"
            >
              Social Studies
            </text>
          </g>
          <g
            transform="translate(190,-120)"
          >
            <rect
              fill="#0984e3"
              height="18"
              width="18"
            />
            <text
              dy="0.35em"
              x="24"
              y="9"
            >
              21st Century Skills
            </text>
          </g>
          <g
            transform="translate(190,-100)"
          >
            <rect
              fill="#e84393"
              height="18"
              width="18"
            />
            <text
              dy="0.35em"
              x="24"
              y="9"
            >
              Arts, Crafts & Labor
            </text>
          </g>
        </g>
      </g>
    </svg>
  </div>
</div>
`;
